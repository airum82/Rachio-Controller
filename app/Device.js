import React, { Component } from 'react';
import ZonesContainer from './ZonesContainer'
import { apiKey } from '../APIkey';
import ControlForm from './ControlForm';

class Device extends Component {
  constructor() {
    super();
    this.state = {
      selectedZones: [],
      zoneIdList: []
    }
    this.runAllZones = this.runAllZones.bind(this);
    this.selectZone = this.selectZone.bind(this);
  }

  runAllZones(duration, event) {
    event.preventDefault()
    const body = {
      zones: this.state.zoneIdList.map((id, index) => {
        return {
          id,
          duration: parseInt(duration),
          sortOrder: index + 1
        }
      })
    }
    console.log(body)
    fetch("https://api.rach.io/1/public/zone/start_multiple", {
      method: 'PUT',
      headers: {
        'Content-Type' : 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify(body)
    })
    .then(response => response.text())
    .then(result => console.log(result))
    .catch(error => console.log(error.message))
  }

  selectZone(id) {
    const isSelected = this.state.selectedZones.some(zone => {
      return zone === id
    })
    if(isSelected) {
      this.setState({
        selectedZones: this.state.selectedZones.filter(zone => {
          return zone !== id
        })
      })
    } else {
      this.setState({ 
        selectedZones: [...this.state.selectedZones, id]
      })
    }
  }

  componentDidMount() {
    this.setState({
      zoneIdList: this.props.zones.map(zone => {
        return zone.id
      })
    })
  }

  render() {
    return (
      <div className="device" id={this.props.id}>
        <h2>{this.props.name}</h2>
        <button onClick={this.props.history.goBack}>back</button>
        <ControlForm runAllZones={this.runAllZones}/>
        <ZonesContainer 
          zones={this.props.zones}
          selectZone={this.selectZone}
        />
      </div>
    )
  }
}

export default Device
