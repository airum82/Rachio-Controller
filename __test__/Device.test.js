import React from 'react';
import Device from '../app/Device';
import { shallow } from 'enzyme';

describe('Device', () => {
  let wrapper;
  const mockProps = {
    history: {
      goBack: jest.fn()
    },
    zones: [{ id: '83-b' }, { id: '74-ty' }],
    id: '58493-a',
    name: 'avenger-7-fold'
  };
  beforeEach(() => {
    wrapper = shallow(<Device
      history={mockProps.history}
      zones={mockProps.zones}
      id={mockProps.id}
      name={mockProps.name}
    />);
  })
  it('should match snapshot upon initial render', () => {
    expect(wrapper).toMatchSnapshot();
  })
  it('runZones should call fetch', () => {
    const mockEvent = {
      preventDefault: jest.fn()
    }
    const mockDuration = '70';
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }));
    wrapper.instance().runZones(mockDuration, mockEvent);
    expect(window.fetch).toHaveBeenCalled();
  })
  it('runZones should setstate with error if unsuccessful', async () => {
    const mockEvent = {
      preventDefault: jest.fn()
    }
    const mockDuration = '70';
    window.fetch = jest.fn().mockImplementation(() => Promise.reject({
      message: 'could not start zones'
    }));
    const spy = await jest.spyOn(wrapper.instance(), 'setState');
    await wrapper.instance().runZones(mockDuration, mockEvent);
    expect(spy).toHaveBeenCalledWith({ error: 'could not start zones' });
  })
  it('select zone should add zone to selected zones if not selected', () => {
    const mockId = '596-a';
    const mockSortOrder = '5'
    wrapper.instance().selectZone(mockId, mockSortOrder);
    expect(wrapper.state().selectedZones[0]).toEqual({
      id: mockId,
      sortOrder: mockSortOrder
    });
  })
  it('selectZone should remove zone from list if there there', () => {
    const mockId = '596-a';
    wrapper.instance().setState({ selectedZones: [{id: '596-a'}] });
    wrapper.instance().selectZone(mockId);
    expect(wrapper.state().selectedZones.length).toEqual(0);
  })
  it('componentDidMount should set state with zone ids', async () => {
    const spy = await jest.spyOn(wrapper.instance(), 'setState');
    wrapper.instance().componentDidMount();
    await wrapper.update();
    expect(spy).toHaveBeenCalledWith({ zoneIdList: ['83-b', '74-ty'] });
  })
  it('back button should call history.goBack', () => {
    wrapper.find('.back-button').simulate('click');
    expect(mockProps.history.goBack).toHaveBeenCalled();
  })
})