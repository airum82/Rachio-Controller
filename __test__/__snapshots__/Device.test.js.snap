// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Device should match snapshot upon initial render 1`] = `
<div
  className="device"
  id="58493-a"
>
  <h2
    className="device-title"
  >
    avenger-7-fold
  </h2>
  <button
    className="back-button"
    onClick={[MockFunction]}
  >
    back
  </button>
  <ControlForm
    runZones={[Function]}
  />
  <ZonesContainer
    addSortOrder={[Function]}
    selectZone={[Function]}
    zones={
      Array [
        Object {
          "id": "83-b",
        },
        Object {
          "id": "74-ty",
        },
      ]
    }
  />
</div>
`;
